plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Future.hs

```diff
@@ -73,230 +151,453 @@ their margin. If either party fails to make a margin payment then the contract
+    --   The value of both margin accounts will be paid to the role that
+    --   *didn't* violate the margin requirement
+
+{-# INLINABLE futureStateMachine #-}
+futureStateMachine
+    :: Future
+    -> FutureOwners
+    -> StateMachine FutureState FutureAction
+futureStateMachine ft fos =
+    StateMachine
+        { smTransition = futureTransition
+        , smCheck      = futureCheck ft fos
+        , smFinal      = \case { Finished -> True; _ -> False }
+        }
 
+{-# INLINABLE mkValidator #-}
+mkValidator :: Future -> FutureOwners -> Scripts.ValidatorType (SM.StateMachine FutureState FutureAction)
+mkValidator future ftos = SM.mkValidator (futureStateMachine future ftos)
+
+scriptInstance :: Future -> FutureOwners -> Scripts.ScriptInstance (SM.StateMachine FutureState FutureAction)
+scriptInstance future ftos = 
+    let val = $$(PlutusTx.compile [|| validatorParam ||])
+            `PlutusTx.applyCode`
+                PlutusTx.liftCode future
+                `PlutusTx.applyCode`
+                    PlutusTx.liftCode ftos
+        validatorParam f g = SM.mkValidator (futureStateMachine f g)
+        wrap = Scripts.wrapValidator @FutureState @FutureAction
+    
+    in Scripts.Validator @(SM.StateMachine FutureState FutureAction)
+        val
+        $$(PlutusTx.compile [|| wrap ||])
+        
+machineInstance
+    :: Future
+    -> FutureOwners
+    -> SM.StateMachineInstance FutureState FutureAction
+machineInstance future ftos = 
+    let machine = futureStateMachine future ftos
+        script  = scriptInstance future ftos
+    in SM.StateMachineInstance machine script
+
+validatorScript :: Future -> FutureOwners -> ValidatorScript
+validatorScript ft fos = Ledger.mkValidatorScript $
+    $$(PlutusTx.compile [|| validatorParam ||])
+        `PlutusTx.applyCode`
+            PlutusTx.liftCode ft
+            `PlutusTx.applyCode`
+                PlutusTx.liftCode fos
+    where validatorParam f g = Scripts.wrapValidator (futureCheck f g)
+    
+{-# INLINABLE verifyOracle #-}
+verifyOracle :: Future -> OracleValue a -> (Slot, a)
+verifyOracle Future{ftPriceOracle} OracleValue{ovSignature, ovSlot, ovValue} =
+    if ovSignature == ftPriceOracle then (ovSlot, ovValue) else error ()
+
+{-# INLINABLE futureTransition #-}
+futureTransition :: FutureState -> FutureAction -> Maybe FutureState
+futureTransition (Running accounts) (AdjustMargin role value) = Just (Running (adjustMargin role value accounts))
+futureTransition (Running accounts) (Settle ov)               = Just Finished
+futureTransition (Running accounts) (SettleEarly ov)          = Just Finished
+futureTransition _ _                                          = Nothing
+
+{-# INLINABLE outputsWith #-}
+-- | The outputs of the 'PendingTx' that carry a non-zero amount of the currency
+--   defined by the 'CurrencySymbol' and the 'TokenName'.
+outputsWith :: PendingTx -> CurrencySymbol -> TokenName -> [PendingTxOut]
```

Comment:
This and other functions for working with specific tokens (i.e. `(CurrencySymbol, TokenName)`) seem like they could live in a library.

Reply:
I put it all into `Ledger.Tokens`